### PR TITLE
PERF: efficient argmax/argmin for SparseArray

### LINKED ISSUE #34197: ENH: implement efficient sorting methods for SparseArray (argsort/argmin/argmax)
xref https://github.com/pandas-dev/pandas/pull/27801#discussion_r426017444

`SparseArray` currently does not implement `_values_for_argsort` or `argsort/`argmin`/`argmax` itself, which means it is using the base implementation, which converts the sparse array to a dense numpy array. Room for improvements thus.

### PR DESCRIPTION
**Pull request by GYHHAHA**
- [ ] partially closes #34197
- [x] [Tests added and passed](https://pandas.pydata.org/pandas-docs/dev/development/contributing_codebase.html#writing-tests) if fixing a bug or adding a new feature
- [x] All [code checks passed](https://pandas.pydata.org/pandas-docs/dev/development/contributing_codebase.html#pre-commit).
- [ ] Added [type annotations](https://pandas.pydata.org/pandas-docs/dev/development/contributing_codebase.html#type-hints) to new arguments/methods/functions.
- [x] Added an entry in the latest `doc/source/whatsnew/vX.X.X.rst` file if fixing a bug or adding a new feature.

Thanks for the review. Currently, only argmax/argmin are implemented since argsort has many annoying corner cases and I have to spent more time on the correctness and take the follow-up in a separate PR. Simple benchmark:
```
>>> val = np.random.rand(1000000)
>>> mask = val < 0.99
>>> val[mask] = np.nan
>>> arr = SparseArray(val)
>>> %timeit arr.argmax()
8.6 ms ± 343 µs per loop (mean ± std. dev. of 7 runs, 100 loops each) <- master
44.3 µs ± 934 ns per loop (mean ± std. dev. of 7 runs, 10,000 loops each) <- this pr
```

### CODE REVIEW

File: pandas/core/arrays/base.py
Comment (mzeitlin11) [inline pandas/core/arrays/base.py:715]: Instead of adding this workaround, what about just directly handling this in `SparseArray` by accessing `_sparse_values`?

File: pandas/core/arrays/base.py
Comment (mzeitlin11) [inline pandas/core/arrays/base.py:671]: Similar comment here, probably better to keep this logic in `SparseArray` instead of adding another method to the `ExtensionArray` interface

File: pandas/core/arrays/sparse/array.py
Comment (mzeitlin11) [inline pandas/core/arrays/sparse/array.py:1187]: See other comments - why not do this directly instead of dispatching `SparseArray -> ExtensionArray -> SparseArray`

File: pandas/tests/arrays/sparse/test_reductions.py
Comment (mzeitlin11) [inline pandas/tests/arrays/sparse/test_reductions.py:291]: Can we also test an empty array input here?

**Review by mzeitlin11 — COMMENTED** [review]
Thanks for the pr @GYHHAHA! Some comments below, general question is why can't we just dispatch to a `SparseArray.argmin/argmax` which directly uses `np.argmin/max` instead of having to modify the general `ExtensionArray` path (for example, how `SparseArray` handles `min/max`. This could also avoid materializing the data.

**Comment by GYHHAHA** [comment]
@mzeitlin11 I think you are right. Originally I want to reuse some general codes. It seems tobe unnecessary. I will handle this in `SparseArray`.

**Comment by mzeitlin11** [comment]
> I find the current implementation for `_first_fill_value_loc` is wrong. A simple bad case:
> 
> ```python
> >>> arr = SparseArray([np.nan, 1, 0, 0, np.nan, 2], fill_value=1)
> >>> arr._first_fill_value_loc()
> 5 # should be 1
> ```
> 
> I will commit a fix after merging this one.

I'm not entirely sure what that function is meant to do, but the name might also just be misleading. It looks to only be used in `unique` as some kind of mechanism for figuring out where to insert the fill value (probably to keep sorted order in the result?).

File: pandas/core/arrays/sparse/array.py
Comment (mzeitlin11) [inline pandas/core/arrays/sparse/array.py:1661]: As you mentioned below, `_first_fill_value_loc` appears to be named confusingly and does more work than needed (involves a sort as well). If you just want to see whether or not the fill value exists, could use `ngaps` or `npoints` on `self.sp_index` (for example as done in `_min_max` or `unique`)

Comment (GYHHAHA) [reply pandas/core/arrays/sparse/array.py:1661]: No, actually the name is fine, but the doc and implementation is wrong. I do need to find the first position of the fill value in argmax/argmin instead of just checking the existence.

Comment (GYHHAHA) [reply pandas/core/arrays/sparse/array.py:1661]: But for the unique function, you are right. We can use O(1) instead O(log(n)) to check the existence.

**Comment by GYHHAHA** [comment]
Apart from the bug mentioned, PERF for unique is a another follow-up to do (latter).

**Comment by GYHHAHA** [comment]
The reason why unique need to find the loc is to keep the insertion place right, which aligns the normal array.unique() result. @mzeitlin11

File: pandas/core/arrays/sparse/array.py
Comment (GYHHAHA) [reply pandas/core/arrays/sparse/array.py:1661]: I can give you the example:
```
>>> arr = SparseArray([1, 2, 0, 0], fill_value=0)
>>> arr.argmin()
2
```
The candidate is 1 but fill_value is smaller, thus we return the result from _first_fill_value_loc().

Comment (mzeitlin11) [reply pandas/core/arrays/sparse/array.py:1661]: Ok thanks for explaining, makes sense. Based on code in `SparseIndex`, for example https://github.com/pandas-dev/pandas/blob/9f43ec6a98580d8f1c74d068afb31ec902c63e8c/pandas/_libs/sparse.pyx#L75, we should be guaranteed monotonically increasing indices. So is the first fill value location just `self.sp_index.indices[0]`?

Comment (GYHHAHA) [reply pandas/core/arrays/sparse/array.py:1661]: `self.sp_index.indices[0]` is actually the first non-fill_value position, not the fill_value.

**Comment by mzeitlin11** [comment]
@GYHHAHA what's the status here? This generally lgtm - but does the `_first_fill_value_loc` issues you discovered cause any regressions in cases where `_first_fill_value_loc` is wrong? Also just confirming benchmarks still look good with new impl?

**Comment by GYHHAHA** [comment]
How about I submit the fix PR for `unique` and `_first_fill_value_loc ` (and add tests) now? After the fix merged, then we can deal with the current one. @mzeitlin11

**Comment by mzeitlin11** [comment]
> How about I submit the fix PR for `unique` and `_first_fill_value_loc ` (and add tests) now? After the fix merged, then we can deal with the current one. @mzeitlin11

Sounds great, thanks!

File: pandas/core/arrays/sparse/array.py
Comment (mzeitlin11) [inline pandas/core/arrays/sparse/array.py:1653]: `.indices` is defined as a property on the `SparseIndex` class, so is not free to calculate. Better to use access once, store in local instead of accessing multiple times

File: pandas/core/arrays/sparse/array.py
Comment (mzeitlin11) [inline pandas/core/arrays/sparse/array.py:1645]: Not necessarily for this PR, since this is already a big improvement, but does this `np.array(isna(self))` cause full array materialization? Could you avoid that by instead just computing `isna` on the sparse values themselves?

**Review by mzeitlin11 — APPROVED** [review]
Thanks @GYHHAHA! LGTM pending one small comment. Can you please also show latest benchmarks for the cases you showed in the pr description?

File: pandas/core/arrays/sparse/array.py
Comment (GYHHAHA) [reply pandas/core/arrays/sparse/array.py]: This is an important point. Thanks.

**Comment by GYHHAHA** [comment]
@mzeitlin11 New benchmark is added in the description.

File: doc/source/whatsnew/v1.5.0.rst
Comment (mroeschke) [inline doc/source/whatsnew/v1.5.0.rst:282]: Could you put this under the `Performance improvements` section?

File: doc/source/whatsnew/v1.5.0.rst
Comment (mroeschke) [inline doc/source/whatsnew/v1.5.0.rst:805]:
```suggestion
- Performance improvement in ``argmax`` and ``argmin`` for :class:`arrays.SparseArray` (:issue:`34197`)
```

**Review by mroeschke — APPROVED** [review]

**Comment by mroeschke** [comment]
Thanks @GYHHAHA

**Pull request merged**
